src/hotspot/share/services/memReporter.cpp

```diff
@@ -820,3 +822,68 @@ void MemDetailDiffReporter::diff_virtual_memory_site(const NativeCallStack* stac
 
   out->print_cr(")\n");
  }
+
+void MemJFRReporter::sendTotalEvent() {
+  if (!MemTracker::enabled()) {
+    return;
+  }
+
+  MemBaseline usage;
+  usage.baseline(true);
+
+  const size_t malloced_memory = usage.malloc_memory_snapshot()->total();
+  const size_t reserved_memory = usage.virtual_memory_snapshot()->total_reserved();
+  const size_t committed_memory = usage.virtual_memory_snapshot()->total_committed();
+
+  const size_t reserved = malloced_memory + reserved_memory;
+  const size_t committed = malloced_memory + committed_memory;
+
+  EventNativeMemoryUsage event;
+  event.set_reserved(reserved);
+  event.set_committed(committed);
+  event.commit();
+}
+
+void MemJFRReporter::sendTypeEvent(const char* type, size_t reserved, size_t committed) {
+  EventNativeMemoryUsagePart event;
+  event.set_type(type);
+  event.set_reserved(reserved);
+  event.set_committed(committed);
+  event.commit();
+}
+
+void MemJFRReporter::sendTypeEvents() {
+  if (!MemTracker::enabled()) {
+    return;
+  }
+
+  MemBaseline usage;
+  usage.baseline(true);
```

Comment:
Perhaps we could collect event timing here, rather than reading the timer for each memory category? That way it’s easier to align events from a single snapshot.

for this sort of event, I’m not sure if it’s preferable to declare only a start time, start==end, or measure start and end time around the ‘baseline(true)’ call.

Reply:
Sounds like a good idea. The only potential problem would be a thread view trying to visualize all event when they were sent and seeing a lot being exactly equal might be strange. I'll consult the JFR guys to see what's best to make it easy to verify which samples are from the same baseline.

Reply:
Sounds good. We have tried to use the same timestamp for periodic events that come together. This simplifies processing of events, since the timestamp can be used as a key.

Looking at other periodic events, they only set period="everyChunk", which I think means startTime=false implicitly. I think it would be good if this follows the pattern.